### PR TITLE
Bluetooth: Add additional attributes to DIS

### PR DESCRIPTION
**Pull request by hersp**
Add SW, HW, FW & Serial attributes to DIS
Add support for reading attributes from settings

Signed-off-by: Peter Herager <pehe@oticon.com>

### CODE REVIEW

File: subsys/bluetooth/services/dis.c [hunk 1]
```diff
@@ -38,6 +47,33 @@ static struct dis_pnp dis_pnp_id = {
...
 
+#ifdef CONFIG_BT_GATT_DIS_DYNAMIC
+#define DIS_STR_MAX CONFIG_BT_GATT_DIS_STR_MAX
+#define DIS_STORAGE /* empty */
+#else
+#define DIS_STR_MAX /* empty */
+#define DIS_STORAGE const
+#endif
+
+static DIS_STORAGE u8_t dis_model[DIS_STR_MAX] = CONFIG_BT_GATT_DIS_MODEL;
```
Comment (Vudentz) [inline subsys/bluetooth/services/dis.c]: I don't get why we need to declare a variable to store the config options? It seems to be we would be wasting memory as one would have to assign CONFIG_BT_GATT_DIS_STR_MAX with the biggest string size.

File: subsys/bluetooth/host/Kconfig [hunk 2]
```diff
@@ -646,6 +646,13 @@ config BT_DEBUG_SDP
 		This option enables debug support for the Bluetooth
 		Service Discovery Protocol (SDP).
 
+config BT_DEBUG_DIS
+	bool "Bluetooth Device Information Service(DIS) debug"
+	depends on BT_GATT_DIS
+	help
+		This option enables debug support for the Bluetooth
+		Device Information Service(DIS).
+
```
Comment (Vudentz) [inline subsys/bluetooth/host/Kconfig]: Id split this change in a separate patch.

File: subsys/bluetooth/services/Kconfig.dis [hunk 3]
```diff
@@ -7,9 +7,24 @@
...
 menuconfig BT_GATT_DIS
 	bool "Enable GATT Device Information service"
+	default n
 
 if BT_GATT_DIS
 
+config BT_GATT_DIS_DYNAMIC
+	bool "Enable Dynamic data in Device Information Service"
+	help
+	 Enable Dynamic data in Device Information Service.
```
Comment (Vudentz) [inline subsys/bluetooth/services/Kconfig.dis]: Im not quite happy about the use of Dynamic term here, it perhaps lead to the impression that it can be changed at runtime or that it is writeable, maybe Options would be better here?

File: subsys/bluetooth/services/dis.c [hunk 4]
```diff
@@ -59,21 +95,145 @@ static ssize_t read_pnp_id(struct bt_conn *conn,
...
+
+			BT_DBG("Software revision set to %s",
+			       dis_software_revision);
+		}
+		return 0;
+	}
+#endif
+}
+
+BT_SETTINGS_DEFINE(dis, dis_set, NULL, NULL);
```
Comment (Vudentz) [inline subsys/bluetooth/services/dis.c]: So this is loading some setting saved on flash but there is no code to store these so I wonder how it gets there in the first place? Furthermore, we do have STR options to set these which gets overwritten if set in flash, Id say we might not need these settings if that can be configured via Kconfig that is probably much easier to setup then have it somewhat stored manually just to be loaded. The general guidance for using the settings subsystem is for values that do change at runtime and need to be persisted, if it is just hardcoded as config option that should not need a setting even if there is a new firmware to be flashed that can just change the config option value.

File: subsys/bluetooth/host/Kconfig (hunk 2 above)
Comment (hersp) [reply subsys/bluetooth/host/Kconfig]: I can do that

File: subsys/bluetooth/services/dis.c (hunk 4 above)
Comment (hersp) [reply subsys/bluetooth/services/dis.c]: In the system I am working on these values can be changed after final compilation as the device could be sold under different brands with each their own versioning scheme. Thus these values will be written in the settings during production and read from setting during boot. so kconfig is not an option.

File: subsys/bluetooth/services/Kconfig.dis (hunk 3 above)
Comment (hersp) [reply subsys/bluetooth/services/Kconfig.dis]: Copied the term from the device name that can be changed runtime. So no problem changing it to settings.

File: subsys/bluetooth/services/dis.c (hunk 1 above)
Comment (hersp) [reply subsys/bluetooth/services/dis.c]: This is only the case when reading from the settings system. Otherwise it is just a const with length defined by the kconfig.

Comment (hersp) [reply subsys/bluetooth/services/dis.c]: Can see the the const concept poses some problems during checking, I will work out something else.

File: subsys/bluetooth/services/dis.c (hunk 4 above)
Comment (Vudentz) [reply subsys/bluetooth/services/dis.c]: > In the system I am working on these values can be changed after final compilation as the device could be sold under different brands with each their own versioning scheme. Thus these values will be written in the settings during production and read from setting during boot. so kconfig is not an option.

So the idea is that it always compiled with the same value which always gets overwritten by what is stored in the settings? Im not sure if there is a practical difference in compiling with a different prj.conf though versus fabricating the settings in flash, afaik for doing the latter there is no tool yet, or there is something on the works? Either way with this design we end up having 3 copies of the same string, one for config option, another stored as setting and yet another for storing the either one of the above, so I can see companies actually favoring rebuilding with a different prj.conf rather than fabricating the flash with settings like you are suggesting. Another question about this design is if by trying to fabricate the flash wouldn't this wipe out existing settings? Afaik to do this properly on an existing image one would have to open the flash fs, read all the stored settings (keys, CCC, etc) and then reconstruct the fs, with the config option alone I beleave this would not be necessary because that doesn't change the flash where the settings are stored.

Comment (hersp) [reply subsys/bluetooth/services/dis.c]: The solution will not prevent compiling with different prj.conf files and using this way of setting up the system but will just provide an alternative way of controlling this for those who needs this. The default one will be the way it is working today.
I do not know of any tools providing being in the works, but e.g. blue address, IRK and device name will be read during boot. They will have to come from somewhere - don't know how this is done.
BTW. In my system the flash(settings) are generated much later in the flow than the compilation of zephyr as it is part of a larger setup.

Comment (Vudentz) [reply subsys/bluetooth/services/dis.c]: Right, though it is still unclear to me how DFU would work with these values stored as settings, or you are saying that we can do partial writes that just overwrites some blocks? Perhaps there should be another Kconfig just to enable the settings loading, that way we only need to enable the extra storage if the settings loading is enabled otherwise you can assign the config options directly as user_data of the attributes.

File: subsys/bluetooth/host/Kconfig (hunk 2 above)
Comment (Vudentz) [reply subsys/bluetooth/host/Kconfig]: I spoke to Johan about this change, we agreed that it better to have just one option which enables debug for every service as oppose to have one per service where we may end up bloating the system with debug options. Also the help indentation is wrong, it seems you have copied it from SDP which is also wrong (it should be just 2 spaces.)

Comment (hersp) [reply subsys/bluetooth/host/Kconfig]: Agree

File: subsys/bluetooth/services/dis.c (hunk 4 above)
Comment (hersp) [reply subsys/bluetooth/services/dis.c]: I changed it so now, when settings are not used it will work as previously using the kconfig.

My knowledge about the DFU is limited, as it is not used in the system I work on.

File: subsys/bluetooth/services/Kconfig.dis (hunk 3 above)
Comment (hersp) [reply subsys/bluetooth/services/Kconfig.dis]: Changed

File: subsys/bluetooth/host/Kconfig (hunk 2 above)
Comment (hersp) [reply subsys/bluetooth/host/Kconfig]: Changed

File: subsys/bluetooth/services/dis.c (hunk 1 above)
Comment (hersp) [reply subsys/bluetooth/services/dis.c]: Now using kconfig directly when not using settings system

File: subsys/bluetooth/services/dis.c [hunk 5]
```diff
@@ -38,6 +47,48 @@ static struct dis_pnp dis_pnp_id = {
 };
 #endif
 
+#ifdef CONFIG_BT_GATT_DIS_SETTINGS
```
Comment (jhedberg) [inline subsys/bluetooth/services/dis.c]: minor: we try to be consistent and use `if defined(CONFIG_BT_...)` throughout the Bluetooth subsystem. It doesn't make much difference for a single condition, but for multiple ones it has the benefit that you can chain them with `||` or `&&` (something you can't do with `#ifdef CONFIG_BT...`).

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/services/dis.c (hunk 5 above)
Comment (hersp) [reply subsys/bluetooth/services/dis.c]: OK - Changed - Did not know.

File: subsys/bluetooth/services/Kconfig.dis [hunk 6]
```diff
@@ -7,9 +7,24 @@
 
 menuconfig BT_GATT_DIS
 	bool "Enable GATT Device Information service"
+	default n
```
Comment (jhedberg) [inline subsys/bluetooth/services/Kconfig.dis]: `n` is the implicit default, so an explicit `default n` should never be used (sorry for not noticing this earlier and causing these extra review rounds)

**Review by jhedberg — CHANGES_REQUESTED** [review]
One more minor thing. Also, please be consistent with capital `B` for `Bluetooth` in the commit message. We just discussed this on #bluetooth on slack today.

File: subsys/bluetooth/services/Kconfig.dis (hunk 6 above)
Comment (hersp) [reply subsys/bluetooth/services/Kconfig.dis]: Done

**Review by Vudentz — APPROVED** [review]
LGTM

File: subsys/bluetooth/services/Kconfig.dis [hunk 7]
```diff
@@ -85,4 +99,47 @@ config BT_GATT_DIS_PNP_VER
...
+	help
+	 Enable Serial Number characteristic in Device Information Service.
+
+config BT_GATT_DIS_SERIAL_NUMBER_STR
+	string "Serial Number"
+	depends on BT_GATT_DIS_SERIAL_NUMBER
+	help
+	 Enable Serial Number characteristic in Device Information Service.
+
+config BT_GATT_DIS_FIRMWARE_REVISION
```
Comment (carlescufi) [inline subsys/bluetooth/services/Kconfig.dis]: Can we make those shorter by using `FW`, `HW` and `SW` instead of `FIRMWARE`, `HARDWARE` and `SOFTWARE`?

File: subsys/bluetooth/services/dis.c [hunk 8]
```diff
@@ -38,6 +47,54 @@ static struct dis_pnp dis_pnp_id = {
...
+#define BT_GATT_DIS_MANUF_REF			dis_manuf
+#define BT_GATT_DIS_SERIAL_NUMBER_STR_REF	dis_serial_number
+#define BT_GATT_DIS_FIRMWARE_REVISION_STR_REF	dis_firmware_revision
+#define BT_GATT_DIS_HARDWARE_REVISION_STR_REF	dis_hardware_revision
+#define BT_GATT_DIS_SOFTWARE_REVISION_STR_REF	dis_software_revision
+
+#else /* CONFIG_BT_GATT_DIS_SETTINGS */
+
+#define BT_GATT_DIS_MODEL_REF			\
+		CONFIG_BT_GATT_DIS_MODEL
```
Comment (carlescufi) [inline subsys/bluetooth/services/dis.c]: I wonder if we should cast to `const` to make sure those strings are stored in ROM and not RAM

**Review by carlescufi — COMMENTED** [review]
Thanks for this welcome change! Added a couple of minor comments

File: subsys/bluetooth/services/Kconfig.dis (hunk 7 above)
Comment (hersp) [reply subsys/bluetooth/services/Kconfig.dis]: Can do that

File: subsys/bluetooth/services/dis.c (hunk 8 above)
Comment (hersp) [reply subsys/bluetooth/services/dis.c]: Not sure what to do here. You cannot just put the string into a const variable as there is no guarantee that you can reference a const variable in ROM as a pointer(depending on compiler/processor).

**Review by carlescufi — APPROVED** [review]

File: subsys/bluetooth/services/dis.c [hunk 9]
```diff
@@ -38,6 +47,46 @@ static struct dis_pnp dis_pnp_id = {
 };
 #endif
 
+#if defined(CONFIG_BT_GATT_DIS_SETTINGS)
+#define DIS_STR_MAX
+static u8_t dis_model[CONFIG_BT_GATT_DIS_STR_MAX] = CONFIG_BT_GATT_DIS_MODEL;
```
Comment (lemrey) [inline subsys/bluetooth/services/dis.c]: All these can be `const`.

File: subsys/bluetooth/services/dis.c [hunk 10]
```diff
@@ -38,6 +47,46 @@ static struct dis_pnp dis_pnp_id = {
 };
 #endif
 
+#if defined(CONFIG_BT_GATT_DIS_SETTINGS)
+#define DIS_STR_MAX
```
Comment (lemrey) [inline subsys/bluetooth/services/dis.c]: Where is this used?

**Review by lemrey — APPROVED** [review]
LGTM, left two minor comments.

File: subsys/bluetooth/services/dis.c (hunk 9 above)
Comment (hersp) [reply subsys/bluetooth/services/dis.c]: It is not consts and also you cannot take the address of a const in all systems.

File: subsys/bluetooth/services/dis.c (hunk 10 above)
Comment (hersp) [reply subsys/bluetooth/services/dis.c]: Oups! Remains from previous version. Will be removed

File: subsys/bluetooth/services/dis.c [hunk 11]
```diff
@@ -59,15 +107,133 @@ static ssize_t read_pnp_id(struct bt_conn *conn,
...
+#if defined(CONFIG_BT_GATT_DIS_SW_REV)
+	BT_GATT_CHARACTERISTIC(BT_UUID_DIS_SOFTWARE_REVISION,
+			       BT_GATT_CHRC_READ, BT_GATT_PERM_READ,
+			       read_str, NULL, BT_GATT_DIS_SW_REV_STR_REF),
+#endif
+
 );
+
+#if defined(CONFIG_BT_SETTINGS) && defined(CONFIG_BT_GATT_DIS_SETTINGS)
+static int dis_set(int argc, char **argv, size_t len_rd, settings_read_cb read,
```
Comment (joerchan) [inline subsys/bluetooth/services/dis.c]: rename read argument to read_cb, to avoid confusion?

File: samples/bluetooth/peripheral_dis/src/main.c [hunk 12]
```diff
@@ -44,6 +45,69 @@ static struct bt_conn_cb conn_callbacks = {
...
+	bt_addr_le_t addr;
+
+	addr.type = BT_ADDR_LE_PUBLIC;
+
+	static u8_t addr_val[6] = { 1, 2, 3, 4, 5, 6 };
+	static u8_t irk[16] = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9,
+				0xa, 0xb, 0xc, 0xd, 0xe, 0xf };
+
+	memcpy(addr.a.val, addr_val, sizeof(addr_val));
+	settings_runtime_set("bt/id", &addr, sizeof(bt_addr_le_t));
```
Comment (joerchan) [inline samples/bluetooth/peripheral_dis/src/main.c]: Any good reason for doing this? This should be removed in my opinion.

**Review by joerchan — CHANGES_REQUESTED** [review]
Looks good. Some minor comments.

File: subsys/bluetooth/services/dis.c (hunk 11 above)
Comment (hersp) [reply subsys/bluetooth/services/dis.c]: Done

File: samples/bluetooth/peripheral_dis/src/main.c (hunk 12 above)
Comment (hersp) [reply samples/bluetooth/peripheral_dis/src/main.c]: As settings are enabled now this will provide the a backend for reading.

Comment (joerchan) [reply samples/bluetooth/peripheral_dis/src/main.c]: But why do you need to store an IRK and Identity Address in settings? That part I don't understand.

**Review by joerchan — APPROVED** [review]

File: samples/bluetooth/peripheral_dis/src/main.c (hunk 12 above)
Comment (jhedberg) [reply samples/bluetooth/peripheral_dis/src/main.c]: This is also unclear to me. Furthermore, we don't support storing/restoring of a public address - only static random address, so this would fail later either way.

Comment (hersp) [reply samples/bluetooth/peripheral_dis/src/main.c]: Agreed. I will remove the non DIS parts.

File: samples/bluetooth/peripheral_dis/src/main.c [hunk 13]
```diff
@@ -44,6 +45,57 @@ static struct bt_conn_cb conn_callbacks = {
 	.disconnected = disconnected,
 };
 
+
```
Comment (jhedberg) [inline samples/bluetooth/peripheral_dis/src/main.c]: Redundant empty line (there should never be a need to have two consecutive ones)

File: subsys/bluetooth/services/dis.c [hunk 14]
```diff
@@ -59,15 +107,134 @@ static ssize_t read_pnp_id(struct bt_conn *conn,
...
+		return 0;
+	}
+#endif
+	return 0;
+}
+
+BT_SETTINGS_DEFINE(dis, dis_set, NULL, NULL);
+#endif /* CONFIG_BT_GATT_DIS_SETTINGS && CONFIG_BT_SETTINGS*/
+
+
```
Comment (jhedberg) [inline subsys/bluetooth/services/dis.c]: Two unnecessary empty lines at end of file, please remove

**Review by jhedberg — CHANGES_REQUESTED** [review]
This looks generally good to me, with the exception of some redundant whitespace (empty lines). Could you fix those, and then we can merge this.

File: samples/bluetooth/peripheral_dis/src/main.c (hunk 13 above)
Comment (hersp) [reply samples/bluetooth/peripheral_dis/src/main.c]: OK

File: subsys/bluetooth/services/dis.c (hunk 14 above)
Comment (hersp) [reply subsys/bluetooth/services/dis.c]: OK

File: subsys/bluetooth/services/dis.c [hunk 15]
```diff
@@ -38,6 +47,45 @@ static struct dis_pnp dis_pnp_id = {
...
+#define BT_GATT_DIS_MODEL_REF		    CONFIG_BT_GATT_DIS_MODEL
+#define BT_GATT_DIS_MANUF_REF		    CONFIG_BT_GATT_DIS_MANUF
+#define BT_GATT_DIS_SERIAL_NUMBER_STR_REF   CONFIG_BT_GATT_DIS_SERIAL_NUMBER_STR
+#define BT_GATT_DIS_FW_REV_STR_REF	    CONFIG_BT_GATT_DIS_FW_REV_STR
+#define BT_GATT_DIS_HW_REV_STR_REF	    CONFIG_BT_GATT_DIS_HW_REV_STR
+#define BT_GATT_DIS_SW_REV_STR_REF	    CONFIG_BT_GATT_DIS_SW_REV_STR
+
+#endif /* CONFIG_BT_GATT_DIS_SETTINGS */
+
+
```
Comment (hersp) [reply subsys/bluetooth/services/dis.c]: OK

File: samples/bluetooth/peripheral_dis/prj.conf [hunk 16]
```diff
@@ -1,5 +1,28 @@
 CONFIG_BT=y
 CONFIG_BT_PERIPHERAL=y
+
+
```
Comment (hersp) [reply samples/bluetooth/peripheral_dis/prj.conf]: OK

File: samples/bluetooth/peripheral_dis/prj.conf [hunk 17]
```diff
@@ -1,5 +1,28 @@
...
+# Below is setup to let DIS information be read from settings
+CONFIG_SETTINGS_RUNTIME=y
+CONFIG_SETTINGS_CUSTOM=y
+CONFIG_SETTINGS=y
+CONFIG_BT_SETTINGS=y
+
+CONFIG_BT_GATT_DIS_SETTINGS=y
+CONFIG_BT_GATT_DIS_STR_MAX=21
+
```
Comment (hersp) [reply samples/bluetooth/peripheral_dis/prj.conf]: OK

**Review by jhedberg — APPROVED** [review]

File: subsys/bluetooth/services/dis.c (hunk 15 above)
Comment (jhedberg) [inline subsys/bluetooth/services/dis.c]: Redundant empty line

File: samples/bluetooth/peripheral_dis/prj.conf (hunk 17 above)
Comment (jhedberg) [inline samples/bluetooth/peripheral_dis/prj.conf]: Extra empty line at the end of the file, please remove

File: samples/bluetooth/peripheral_dis/prj.conf (hunk 16 above)
Comment (jhedberg) [inline samples/bluetooth/peripheral_dis/prj.conf]: No need for two empty lines, just use one
